stubs/EloquentCollection.stub

```diff
@@ -3,7 +3,7 @@
 namespace Illuminate\Database\Eloquent;
 
 /**
- * @template TModel of Model
+ * @template TModel
```

Comment:
In which case would this not be true? I get that an EloquentCollection can be downgraded to a SupportCollection by transforming it's elements to something that is not a model. However, in that case it would no longer be an EloquentCollection.

Reply:
Although it does not make a lot of sense, this is true quite often. There are two examples in the description of this MR and there is one more in the linked issue.

I really think it's something that should be fixed on Laravel's side as it's quite non-intuitive atm. However, this would be quite a breaking change so I don't see that happening any soon.

Reply:
There are ways that make it possible to put non-models in an EloquentCollection, agreed. That does not make it correct to do so. The goal of static analysis tools is not to validate nonsensical or plain wrong code, i would argue that it is correct for larastan to fail in such cases.

Reply:
But these examples aren't nonsensical or wrong, would you not agree? This is a perfectly normal piece of code:
```php
return User::all()->partition(function (User $user) {
        return $user->accounts()->count() > 1;
})->map(function (EloquentCollection $users) {
        return $users->count();
});
```
If we enforce that an EloquentCollection can only hold instances of a `Model` then Larastan will always complain about this and similar pieces of code. I do not see how we can tackle this problem in a different way without completely dropping support for `partition`, `mapToGroups` and a few other methods.

Reply:
Maybe we could do something like `TModel of Model|EloquentCollection` but it seems this is not supported by PHPStan. https://phpstan.org/r/be88b35a-0f5e-4da4-ae40-93bac1e949b1

Reply:
That piece of code looks fine and quite idiomatic indeed, i 100 % agree that it should be possible to write it this way.

The EloquentCollection already overrides a few functions to return a SupportCollection instead, e.g.:

```php
    /**
     * Flip the items in the collection.
     *
     * @return \Illuminate\Support\Collection
     */
    public function flip()
    {
        return $this->toBase()->flip();
    }
```

Maybe we can try and PR additional methods onto there that also change the elements within the collection?

Until then, i am not too sure what the best way of going forward is. Your PR allows more valid programs, but also fails to detect some incorrect ones.

Reply:
Code such as the one in your example could be fixed like this, until Laravel actually fixes the EloquentCollection:

```php
return User::all()->partition(function (User $user) {
        return $user->accounts()->count() > 1;
})->toBase()->map(function (EloquentCollection $users) {
        return $users->count();
});
```

Reply:
@canvural That would still not be sufficient, as the EloquentCollection can literally hold anything. This can happen when using `mapToGroups` for example.

@spawnia Yeah it'd be perfect if it could be fixed on Laravel's side, it would then become a lot easier on our side. Though they might consider this breaking changes which means we'd still have to wait quite a while. 

I'm fine with not introducing most of these stubs and just waiting / hoping for Laravel to change their return types. However, I do think the current false positives should be fixed asap. Because apart from the false positive example in the issue, there are also false positives when you don't specify the generic type of your Collection which is quite annoying.

Reply:
See https://github.com/laravel/framework/pull/32677 😉

Reply:
@spawnia Yes, I agree. Ideally, this should be handled in Laravel. But currently removing this piece of code prevents a lot of false positives. So I'm gonna merge this right now. But definelty we should discuss/improve on this in a new issue or PR!